rp2040-hal/src/usb.rs

```diff
@@ -1,3 +1,528 @@
+        interrupt::free(|cs| {
+            let inner = self.inner.borrow(cs).borrow_mut();
+            inner
+                .ctrl_reg
+                .addr_endp
+                .modify(|_, w| unsafe { w.address().bits(addr & 0x3F) });
+            // reset ep0
+            inner.ctrl_dpram.ep_buffer_control[0].modify(|_, w| w.pid_0().set_bit());
+            inner.ctrl_dpram.ep_buffer_control[1].modify(|_, w| w.pid_0().set_bit());
+        })
+    }
+    fn write(&self, ep_addr: EndpointAddress, buf: &[u8]) -> UsbResult<usize> {
+        interrupt::free(|cs| {
+            let mut inner = self.inner.borrow(cs).borrow_mut();
+            inner.ep_write(ep_addr, buf)
+        })
+    }
+    fn read(&self, ep_addr: EndpointAddress, buf: &mut [u8]) -> UsbResult<usize> {
+        interrupt::free(|cs| {
+            let mut inner = self.inner.borrow(cs).borrow_mut();
+            inner.ep_read(ep_addr, buf)
+        })
+    }
+    fn set_stalled(&self, ep_addr: EndpointAddress, stalled: bool) {
+        interrupt::free(|cs| {
+            let inner = self.inner.borrow(cs).borrow_mut();
+
+            if ep_addr.index() == 0 {
+                inner.ctrl_reg.ep_stall_arm.modify(|_, w| {
+                    if ep_addr.is_in() {
+                        w.ep0_in().bit(stalled)
+                    } else {
+                        w.ep0_out().bit(stalled)
+                    }
+                });
+            }
+
+            let index = ep_addr_to_ep_buf_ctrl_idx(ep_addr);
+            inner.ctrl_dpram.ep_buffer_control[index].modify(|_, w| w.stall().bit(stalled));
+        })
+    }
+    fn is_stalled(&self, ep_addr: EndpointAddress) -> bool {
+        interrupt::free(|cs| {
+            let inner = self.inner.borrow(cs).borrow_mut();
+            let index = ep_addr_to_ep_buf_ctrl_idx(ep_addr);
+            inner.ctrl_dpram.ep_buffer_control[index]
+                .read()
+                .stall()
+                .bit_is_set()
+        })
+    }
+    fn suspend(&self) {
+        todo!()
+    }
+    fn resume(&self) {
+        todo!()
+    }
+    fn poll(&self) -> PollResult {
+        interrupt::free(|cs| {
+            let mut inner = self.inner.borrow(cs).borrow_mut();
+            // TODO: check for suspend request
+            // TODO: check for resume request
+
+            // check for bus reset
+            let sie_status = inner.ctrl_reg.sie_status.read();
+            if sie_status.bus_reset().bit_is_set() {
+                return PollResult::Reset;
+            }
+
+            let (mut ep_out, mut ep_in_complete, mut ep_setup): (u16, u16, u16) = (0, 0, 0);
+
+            let buff_status = inner.ctrl_reg.buff_status.read().bits();
+            if buff_status != 0 {
+                // IN Complete shall only be reported once.
+                inner
+                    .ctrl_reg
+                    .buff_status
+                    .write(|w| unsafe { w.bits(0x5555_5555) });
+
+                for i in 0..32u32 {
```

Comment:
maybe 2 separate ranges here, 1 that does ep_in_complete and the other that does ep_out.
have them step 2.
I don't think it will produce better assembly but I think it reads a little easier.
```rust
for i in (0..32u32).step_by(2) {
  let m = 1 << i;
  ep_in_complete |= ((buff_status & m) >> (i / 2)) as u16;
}
for i in (1..32u32).step_by(2) {
  let m = 1 << i;
  ep_out |= ((buff_status & m) >> (i / 2 + 1)) as u16;
}
````

Reply:
Because this code is run very often and do not actually contribute much to the usb stack processing I wanted to make it run as fast as possible (and avoid multiple loops).
I agree that it is convoluted and I don't really like it either.

I tried to compare your suggestion with the existing solution and a few alternative.

```rust
// Alternative 1:
while buff_status != 0 {
    let first_bit = buff_status.trailing_zeros();
    
    let u32_mask = 1 << first_bit;
    let u16_mask = 1 << (first_bit / 2);
    
    // clear the flag
    buff_status ^= u32_mask;
    // assign to the appropriate direction.
    if (first_bit & 1) == 0 {
        ep_in_complete |= u16_mask;
    } else {
        ep_out |= u16_mask;
    }
}
```
and
```rust
// Alternative 2
for i in 0..32u32 {
    let mask = 1 << i;
    if (buff_status & mask) == mask {
        // Even index are IN endpoints.
        if (i & 1) == 0 {
            ep_in_complete |= 1 << (i / 2);
        } else {
            ep_out |= 1 << (i / 2);
        }
    }
}
```

Here's what `cargo size --release --example pico_usb_serial` reports:

| snippet | size (bytes) | comment |
| --- | --- | --- |
| current | 29248 | hardly readable
| 9names' suggestion |  29168 | enhance readability but reusing the convoluted bit shifting/masking or the previous solution. |
| alternative 1 | 29168 | Using a ctz that do not exist as an instruction on cm0+. |
| alternative 2 | 29088 | Seems to be the best solution, easier to read and also smaller! |

Do you think the Alternative 2 would do?

Reply:
alt2 does read better.
Is it functionally equivalent to the original code though, or was the +1 in the ep_out shift a bug in the original?

It's hard for anyone else to determine if you fixed or introduced a bug in the fix, as the commit containing the "current" version has been replaced - please just add more commits to your PR once a review has started, it makes it easier to determine what you've changed.
We can squash them when merging to remove fix-up commits.
Here's a compiler explorer version of the above versions + the original version (from my archives) for anyone else that wants to review the code + the performance
https://rust.godbolt.org/z/qMKxdTfWK

current and alt2 produce very similar code, alt1 produces good code for thumbv7 but terrible for thumbv6, my suggestion was significantly worse than all the others.
So alt2 (assuming it's still correct) is a good option.

Reply:
Have not looked at the context here, but would using the ROM `ctz` be an option here? https://github.com/rp-rs/rp-hal/blob/main/rp2040-hal/src/rom_data.rs#L65

Reply:
I don't think we have enough of a performance problem here to justify making this more complex.

Even if we did: In the worst-case (every endpoint needs handling) 32 calls to the ROM ctz (11 or 10 cycles each depending on ROM version) is going to be slower than 32 calls to check if the current bit is zero.

If you were looking for quick gains in the common case, I think using clz to set the end of the range would be better
since the number of endpoints used by most people will be very low.
It does remove the fixed loop size though, which the optimizer can use to make the loop faster, so you are punishing the worst case if you do this

Reply:
>  Is it functionally equivalent to the original code though, or was the +1 in the ep_out shift a bug in the original?

It is functionally equivalent. The +1 was not a bug either. The goal is to map the bits from the position `i` in `buff_status` to the position `i/2` in `ep_out`, so the original version achieved that by moving the bit to the left by `n+1` (regardless if this bit was 0 or 1).
In alt2 we checked if that bit was one so we directly shift 1 to the right by `i/2` and or it into `ep_out`.

> It's hard for anyone else to determine if you fixed or introduced a bug in the fix, as the commit containing the "current" version has been replaced

True, sorry about that. I got used to Gitlab that handles that better.

> would using the ROM `ctz` be an option here?

I had a glance at it but AFAIU the lookup function would have to be every time and I haven't looked at what's in there.

Reply:
My tired brain missed that it was shifted from the mask position, and that's the source of the +1. makes sense.
This is much better though so I'm going to mark this as resolved